Operator level versioning for mobile

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52622 Operator level versioning for mobile**

PR1 of two PRs to support operator-level versioning in mobile. For design details please refer to: https://docs.google.com/document/d/1BNQxUmn8uBKxaRhGyrMn2oI-HC07l7hJXwbnwB9ZiYM/edit 

 At top level, with a tuple of `(operator_name, op_version_in_model,
 bytecode_version_in_model)`, 
check the compatibility for that operator. If
 it's not compatible, throw an error with reason. If it's compatible, do
 compatibility treatments if necessary, and return the corresponding operator
 functor that is compatible to this runtime. The logic is implemented in
 function `operator_resolver()`.

 Under the hood, a static table,
 `op_version_table`, is saved for current runtime. The keys are the operator
 names and values are a container with version numbers and their metadata. The
 incoming operator with version is searched in the table to find the match.
 Note that the size of `op_version_table` is small because it only contains
 operators that have more than one versions. For most of the operators, they
 fall through to the existing implementation with default version 0.

Query API:
The runtime operator table can be queried by `get_op_version_table`. 
It returns a map with key: operator name, val: a pair of supported version range[min, max]
It can be used in deployment time to help compatibility check of models to be deployed.

Tests:
`LiteInterpreterTest.OpVersionTable`

In the next PR, an operator version table will be added to bytecode of the model file, to finish the loop of operator versioning check.

Differential Revision: [D26877408](https://our.internmc.facebook.com/intern/diff/D26877408)

iseeyuan committed:
Operator level versioning for mobile

[ghstack-poisoned]

iseeyuan committed:
Update on "Operator level versioning for mobile"




[ghstack-poisoned]

iseeyuan committed:
Update on "Operator level versioning for mobile"




[ghstack-poisoned]

iseeyuan committed:
Update on "Operator level versioning for mobile"


PR1 of two PRs to support operator-level versioning in mobile. For design details please refer to: https://docs.google.com/document/d/1BNQxUmn8uBKxaRhGyrMn2oI-HC07l7hJXwbnwB9ZiYM/edit 

 At top level, with a tuple of `(operator_name, op_version_in_model,
 bytecode_version_in_model)`, 
check the compatibility for that operator. If
 it's not compatible, throw an error with reason. If it's compatible, do
 compatibility treatments if necessary, and return the corresponding operator
 functor that is compatible to this runtime. The logic is implemented in
 function `operator_resolver()`.

 Under the hood, a static table,
 `op_version_table`, is saved for current runtime. The keys are the operator
 names and values are a container with version numbers and their metadata. The
 incoming operator with version is searched in the table to find the match.
 Note that the size of `op_version_table` is small because it only contains
 operators that have more than one versions. For most of the operators, they
 fall through to the existing implementation with default version 0.

Tests:
`LiteInterpreterTest.OpVersionTable`


[ghstack-poisoned]

iseeyuan committed:
Update on "Operator level versioning for mobile"


PR1 of two PRs to support operator-level versioning in mobile. For design details please refer to: https://docs.google.com/document/d/1BNQxUmn8uBKxaRhGyrMn2oI-HC07l7hJXwbnwB9ZiYM/edit 

 At top level, with a tuple of `(operator_name, op_version_in_model,
 bytecode_version_in_model)`, 
check the compatibility for that operator. If
 it's not compatible, throw an error with reason. If it's compatible, do
 compatibility treatments if necessary, and return the corresponding operator
 functor that is compatible to this runtime. The logic is implemented in
 function `operator_resolver()`.

 Under the hood, a static table,
 `op_version_table`, is saved for current runtime. The keys are the operator
 names and values are a container with version numbers and their metadata. The
 incoming operator with version is searched in the table to find the match.
 Note that the size of `op_version_table` is small because it only contains
 operators that have more than one versions. For most of the operators, they
 fall through to the existing implementation with default version 0.

Tests:
`LiteInterpreterTest.OpVersionTable`

In the next PR, an operator version table will be added to bytecode of the model file, to finish the loop of operator versioning check.

[ghstack-poisoned]

cccclai commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,185 @@
+//   * Route to another op name
+//   * Push additional default values
+// It’s fine to update the content of version info, to handle more situations in
+// fugure. Since it’s implementation details, it should not affect
```
future?

cccclai commented on torch/csrc/jit/mobile/versioned_operators.h:
```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/csrc/jit/api/module.h>
```
Do they need to be in header file?

iseeyuan replied:
Will move to cpp file

cccclai commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,185 @@
+    }
+  }
+
+  if (op_version == container.max_version() + 1) {
```
Hmm why current version is max_version() + 1?

iseeyuan replied:
Good catch! There should be no "+1" here.

mruberry replied:
This is interesting. So if an operation, like convolution, changes, then it will require "dummy" version infos forever so that something in the vector (or at the end of the vector) matches these checks? It also seems like operators that have had multiple changes won't be supported (although this framework could be extended later). 

For example, if there's a function foo(), and over time two arguments are added to foo(), how would that work? So let's say that version 1 of the runtime adds an argument 'a' to foo(), and version 3 of the runtime adds an argument 'b' to foo(). With this approach I don't think foo() from version 0 can be run on version 3 or later, right? And foo() would require an entry for version 1 that adds 'b', an entry for version 2 that adds 'b' and no entry for version 3 as long as the runtime version is also 3. But when the runtime version becomes 4 then foo() will also need an entry for version 3, but that entry won't do anything?

Is that correct?

iseeyuan replied:
Very good point! The table can only support one version that add one arg. If there is another version adding another arg, it won't work. One option is to have a vector of default args (IValues) to the operator version table. In your example, when runtime is on v1, the table would be like {"foo": {0: ["a"]}}. When runtime is on v2, the table is {"foo": {0: ["a"]}, {1: []}} (neglecting changes other than adding args). When runtime is on v3, the table is {"foo": {0: ["a", "b"]}, {1: ["b"]}, {2: ["b"]}}. Would this work?

mruberry replied:
It might be interesting to consider an approach like the operator versioning implemented for torchscript. You could, for example, have each op entry have a start version and end version. That way when loading from version X, you would scan through the former versions and find which op entry X is in, or not discover it and know that no additional processing was needed. 

Sticking with only being able to add a single arg seems OK for now. If you need more advanced conversions in the future I think you can write them later.

```
struct VersionInfo {
  VersionInfo(
    const int64_t _start_version,
    const int64_t _end_version,
    const c10::OperatorName& _name,
    const c10::optional<IValue>& _default_val) 
      : start_version_{_start_version}, end_version_{_end_version}, name_{_name}, default_val_{_default_val} { }

  ...
};
```

So when getting an operator foo with version X, the code can enumerate the VersionInfos like today but check their start and end versions:

```
for (const auto& vi : version_infos) {
  if (version >= vi.start_version_ && version <= vi.end_version_) {
    // match!
    ...
  }
}

// no match found! that means no compatibility is needed and the op can be run normally
```

What do you think of an approach like this? I think it might simplify things.

iseeyuan replied:
I get the point of keeping _start_version and _end_version in each entry. It's especially useful for torchscript versioning, where the op version is bond to the global model version. In such case, even there's no change in op_A, but since op_B is changed, we need to bump _end_version of op_A as well. 

In this PR we decoupled the global model version, and make op version unique for each op. In such case, each entry must be different for different op version. i.e., if op_A is compatibility break, we must add another entry with new version to op_A, otherwise it's not "break". Keeping _start_version and _end_version in each entry may not be practical. 

And out range (no match in your example) does not necessarily mean the op can be run normally. If the op version X is too old that we decide from some time we will not support versions older than X for this op, we should throw and error instead of run it normally.

mruberry replied:
I'm not sure I follow. Maybe a VC would be more helpful than Github to discuss this issue.

What I don't understand, in particular, is:

- I'm not sure why start and end can't refer to op versions
- if an Op from version X shouldn't be run then it seems like that could be modeled as a 0 start X end entry that says to throw a runtime error

iseeyuan replied:
Yeah a VC would make it more efficient 😄 
I'm not saying start and end can't refer to op versions. Just saying they may not be necessary because each time we bump a version is for a breaking reason. i.e. start == end

cccclai commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,185 @@
+  }
+
+  int64_t max_version() const {
+    return version_vec.rbegin()->version + 1;
```
Wny is there a +1 here?

iseeyuan replied:
The table saves the older versions and their compatibility actions. For example, if an op has different treatments for version from 0 to 2, this runtime could also handle 3, which is the default version. For default version, it's like all other operators not in the table. There's no special handling so no need to save here.

mruberry commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,185 @@
+    const c10::OperatorName& opname,
+    int64_t op_version,
+    int64_t model_version) {
+  if (model_version > 0x3LL) {
```
What's up with this magic number? The linter will probably complain that this is not a preprocessor define or constexpr.

iseeyuan replied:
I'll put it in a variable. It's the old bytecode version when we have to make it bc.

mruberry commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,185 @@
+    }
+  } else if (model_version == 0x3LL) {
+    auto fn = findOperatorFromName(opname);
+    if (opname == c10::OperatorName("aten::_convolution", "")) {
```
I'm confused why this special case for convolution is needed here. Shouldn't its entry in op_version_table take care of its handling?

iseeyuan replied:
Models with version 3 does not have the operator version table. Since we can't get op version info from the model, we have to do some manual check.

mruberry replied:
Would you elaborate on this?

If this code is reading an older model then doesn't it still have access to the operator version table because the table is part of the reader's code base?

raziel replied:
I think what Martin means is that for this versioning mechanism to work there needs to be a version associated to the operators in the model as well (to be implemented in a subsequent PR).

Since that's not going to be present in old models, then he needs to fallback. to the old "backward compatibility logic".

mruberry commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,185 @@
+}
+
+TORCH_API std::unordered_map<std::string, std::pair<int64_t, int64_t>>
+get_op_version_table() {
```
Seems like this function could just return a pointer to the actual table?

iseeyuan replied:
This function is a public API to query "the operator versioned with there [min, max] range". It's used for compatibility check. Since it does not need the details (such as default values) that the actual table contains, we provide this simpler return type.

mruberry commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,185 @@
+  // The situation where op_version is in [min_version, max_version + 1], but
+  // not in the container, which should not happen.
+  TORCH_CHECK(
+      true,
```
true -> false so this throws every time

iseeyuan replied:
Good catch!

mruberry commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,185 @@
+      op_version,
+      " is not compatible in this runtime. ");
+
+  return VersionInfo(
```
I think you can eliminate this return once the above check always starts throwing.

raziel commented:
Adding a reminder from our conversation:

1. Let's define concrete boundaries for the up-versioning patterns that we know we can implement (i.e. pushing default values, and calling other operators). That will make it easier for people to implement their specific transformations when op version bump is needed; plus also avoid mistakes. In the future, we can add more to the interface as we identify more patterns.
2. We need a specific field for the current op version. The list of up-version transforms is not enough since there's cases where a version bump can result in up-versioning not being possible, and thus the clearing of the up-version transforms.

iseeyuan committed:
Update on "Operator level versioning for mobile"


PR1 of two PRs to support operator-level versioning in mobile. For design details please refer to: https://docs.google.com/document/d/1BNQxUmn8uBKxaRhGyrMn2oI-HC07l7hJXwbnwB9ZiYM/edit 

 At top level, with a tuple of `(operator_name, op_version_in_model,
 bytecode_version_in_model)`, 
check the compatibility for that operator. If
 it's not compatible, throw an error with reason. If it's compatible, do
 compatibility treatments if necessary, and return the corresponding operator
 functor that is compatible to this runtime. The logic is implemented in
 function `operator_resolver()`.

 Under the hood, a static table,
 `op_version_table`, is saved for current runtime. The keys are the operator
 names and values are a container with version numbers and their metadata. The
 incoming operator with version is searched in the table to find the match.
 Note that the size of `op_version_table` is small because it only contains
 operators that have more than one versions. For most of the operators, they
 fall through to the existing implementation with default version 0.

Query API:
The runtime operator table can be queried by `get_op_version_table`. 
It returns a map with key: operator name, val: a pair of supported version range[min, max]
It can be used in deployment time to help compatibility check of models to be deployed.

Tests:
`LiteInterpreterTest.OpVersionTable`

In the next PR, an operator version table will be added to bytecode of the model file, to finish the loop of operator versioning check.

Differential Revision: [D26877408](https://our.internmc.facebook.com/intern/diff/D26877408)

[ghstack-poisoned]

raziel commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,169 @@
+//   * Push additional default values
+// It’s fine to update the content of version info, to handle more situations in
+// future. Since it’s implementation details, it should not affect compatibility.
+struct VersionInfo {
```
Not sure VersionInfo conveys the meaning.

Maybe VersionResolution, VersionCompatibility, ...?

raziel commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,169 @@
+OperatorFunctor operator_resolver(
+    const c10::OperatorName& opname,
+    int64_t op_version,
+    int64_t model_version) {
```
I see why you needed the model version here. So it's not that we do anything in the new versioning mechanism with the bytecode version; it's just to keep BC with the old way we versioned models, before this new mechanism is implemented.
Basically if we didn't care about about model_version 0x3LL we wouldn't need this parameter at all here.

iseeyuan replied:
Right. It's for backward compatibility. When there's no model_version 0x3LL we could remove the model version part here.

raziel reviewed (commented):
Thanks Martin.

iseeyuan commented on torch/csrc/jit/mobile/versioned_operators.h:
```diff
@@ -0,0 +1,38 @@
+TORCH_API OperatorFunctor operator_resolver(
+    const c10::OperatorName& opname,
+    int64_t op_version,
+    int64_t model_version);
```
It's for backward compatibility for old models, where there's no op version table.

raziel replied:
Ideally we shouldn't have the extra model_version (i.e. bytecode version).
To check whether we can just ignore BC with models < 0x3LL

model_version is only meant to check for model level changes to the bytecode, with the existing min/max version logic (separate from the ops themselves).

iseeyuan commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,169 @@
+// incoming operator with version is searched in the table to find the match.
+// Note that the size of op_version_table is small because it only contains
+// operators with compatibility updates. For most of the operators, they
+// fall through to the existing implementation with default version 0.
```
That's right!

iseeyuan commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,169 @@
+// VersionInfoContainer holds a map of {verion_num: VersionInfo} and an explicit
+// cur_version. cur_version is necessary for the situation where an update
+// invalidate all versions in the container. In such case there's an
+// empty version_vec and the only supported version is cur_version.
```
Good catch. It was for previous code.

iseeyuan commented on torch/csrc/jit/mobile/versioned_operators.h:
```diff
@@ -0,0 +1,38 @@
+ * @return a map with key: operator name, val: a pair of supported version range
+ * [min, max]
+ */
+TORCH_API std::unordered_map<std::string, std::unordered_set<int64_t>>
```
I was using unordered_set here for two reasons. One is that it may be easier for version search for users (like deployment time check). Second I'm not sure if there are situations where "version holes" as @dhruvbird mentioned. Something like the supported versions are {1,3,4} but we don't support 2.

cccclai replied:
Why `int64_t` instead of `uint64_t`?

mruberry reviewed (commented):
Thank you for explaining the updated design to me, @iseeyuan. The issues I was concerned about appear to be addressed.

raziel commented on torch/csrc/jit/mobile/versioned_operators.cpp:
```diff
@@ -0,0 +1,169 @@
+// This file is to handle operator versioning in mobile.
+// How does it work?
+// At top level, with a tuple of (operator_name, op_version_in_model,
+// bytecode_version_in_model), check the compatibility for that operator. If
```
Part of the discussion in the other couple of comments that bytecode_version_in_model is only used for BC under the context of op versioning (bytecode version still has its place outside of op versioning for tracking changes to the bytecode itself).
